### PR TITLE
[FLINK-10415] Fail response future if connection closes in RestClient

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

If the RestClient detects that a connection was closed (channel became inactive), then
it now fails the json response future with a ConnectionClosedException.

## Verifying this change

- Added `RestClientTest#testConnectionClosedHandling` and `RestClientTest#testRestClientClosedHandling`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java [hunk 1]
```diff
@@ -428,6 +447,12 @@ protected void channelRead0(ChannelHandlerContext ctx, Object msg) {
 			ctx.close();
 		}
 
+		@Override
+		public void channelInactive(ChannelHandlerContext ctx) throws Exception {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: doesn't belong to this commit.

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: True, will fix it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java [hunk 2]
```diff
@@ -89,7 +91,78 @@ public void testInvalidVersionRejection() throws Exception {
...
+
+			// establish connection
+			connectionSocket = serverSocket.accept();
+
+			restClient.close();
+
+			try {
+				responseFuture.get();
+			} catch (ExecutionException ee) {
+				assertThat(ExceptionUtils.findThrowable(ee, ConnectionClosedException.class).isPresent(), is(true));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java]: this test is currently failing, and the travis logs aren't that helpful.
We could write the assertion like this to make the error message more helpful:
```
optionalException = ExceptionUtils.findThrowable(ee, ConnectionClosedException.class)
if (!optionalException.isPresent()) {
	throw ee;
}
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java [hunk 3]
```diff
@@ -89,7 +91,78 @@ public void testInvalidVersionRejection() throws Exception {
...
+	 */
+	@Test
+	public void testRestClientClosedHandling() throws Exception {
+		final ServerSocket serverSocket = new ServerSocket(0);
+		final String targetAddress = "localhost";
+		final int targetPort = serverSocket.getLocalPort();
+		Socket connectionSocket = null;
+
+		try (final RestClient restClient = new RestClient(RestClientConfiguration.fromConfiguration(new Configuration()), TestingUtils.defaultExecutor())) {
+			final CompletableFuture<EmptyResponseBody> responseFuture = restClient.sendRequest(
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java]: I'm wondering...consider the `submitRequest` method from the `RestClient`:
```
private <P extends ResponseBody> CompletableFuture<P> submitRequest(String targetAddress, int targetPort, Request httpRequest, JavaType responseType) {
		final ChannelFuture connectFuture = bootstrap.connect(targetAddress, targetPort);

		final CompletableFuture<Channel> channelFuture = new CompletableFuture<>();

		connectFuture.addListener(
			(ChannelFuture future) -> {
				if (future.isSuccess()) {
					channelFuture.complete(future.channel());
				} else {
					channelFuture.completeExceptionally(future.cause());
				}
			});

		return channelFuture
			.thenComposeAsync(
				channel -> {
					ClientHandler handler = channel.pipeline().get(ClientHandler.class);
					CompletableFuture<JsonResponse> future = handler.getJsonFuture();
					try {
						httpRequest.writeTo(channel);
					} catch (IOException e) {
						return FutureUtils.completedExceptionally(new FlinkException("Could not write request.", e));
					}
					return future;
				},
				executor)
			.thenComposeAsync(
				(JsonResponse rawResponse) -> parseResponse(rawResponse, responseType),
executor);
```

What happens if the client is shutdown, before the `thenComposeAsync` portions have been run? Then the channelFuture is completed by the listener; and then subsequent stages aren't executed?

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java]: yes, there were actually a bunch of related problems: Channel inactivity and not initialized pipeline.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java]: I think this situation is fine, because an exceptionally completed future won't trigger the executions of the async stages. Instead it will directly fail the returned future.

**Comment by tillrohrmann** [comment]
Thanks for the first review @zentol. I actually added two more commits which handle the idleness of connections and a problem which occurred in tests that the channel pipeline was not properly initialized. These commits would also need a review.

I'm not 100% sure about the default value for the `IDLENESS_TIMEOUT` (60s).

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestClientTest.java]: but it does fail with a different exception, no? although from a user-perspective the same issue happened.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java [hunk 4]
```diff
@@ -339,12 +338,26 @@ private static Request createRequest(String targetAddress, String targetUrl, Htt
...
 					try {
-						httpRequest.writeTo(channel);
+						if (handler == null) {
+							throw new IOException("Netty pipeline was not properly initialized.");
+						} else {
+							httpRequest.writeTo(channel);
+							future = handler.getJsonFuture();
+							success = true;
+						}
 					} catch (IOException e) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: We should catch all exceptions here, then this issue wouldn't have occurred in the first place.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java [hunk 5]
```diff
@@ -339,12 +338,26 @@ private static Request createRequest(String targetAddress, String targetUrl, Htt
...
 				channel -> {
 					ClientHandler handler = channel.pipeline().get(ClientHandler.class);
-					CompletableFuture<JsonResponse> future = handler.getJsonFuture();
+
+					CompletableFuture<JsonResponse> future;
+					boolean success = false;
+
 					try {
-						httpRequest.writeTo(channel);
+						if (handler == null) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: Do we know when/why this happens?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java [hunk 6]
```diff
@@ -339,12 +338,26 @@ private static Request createRequest(String targetAddress, String targetUrl, Htt
...
+							httpRequest.writeTo(channel);
+							future = handler.getJsonFuture();
+							success = true;
+						}
 					} catch (IOException e) {
-						return FutureUtils.completedExceptionally(new FlinkException("Could not write request.", e));
+						future = FutureUtils.completedExceptionally(new ConnectionException("Could not write request.", e));
+					} finally {
+						if (!success) {
+							channel.close();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: why aren't we closing the channel in both cases? Not being initialized _properly_ doesn't necessarily mean that there's nothing to clean up.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: We retry the send operation if we see an `IOException`. If any other exception occurs (e.g. `RuntimeException`), this might indicate that there is another problem where we don't want to retry. Therefore, I think it is ok to only catch the `IOException`. The request will also be failed if we see a different exception.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: No not really. It happened from time to time when running the `RestClientTest#testConnectionClosedHandling` in a loop. It might be a Netty bug but we should guard against this in order to harden the test case. I don't think that something bad happens if we fail the request with an `IOException` here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java (hunk 6 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: We keep the channel open if we managed to send the request to the remote endpoint, because we expect an answer. In all other cases (`!success`) we should close it, because we cannot expect something to return.

**Comment by tillrohrmann** [comment]
Thanks again for the review @zentol. I had to add the `IdleStateHandler` because it happened sometimes in `RestClientTest#testConnectionClosedHandling` that the channel would not be closed by Netty after we closed the server socket. So sometimes Netty detects such an event and then calls `channelInactive` and sometimes not. For the latter case, we need to monitor the connection for idleness.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java (hunk 6 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java]: ah of course 👍

**Comment by zentol** [comment]
urgh, well that's unfortunate. We may want to increase the idle timeout to a larger value though (maybe 5 minutes?), as it effectively is an upper limit for the timeout that a user may specify.

**Comment by tillrohrmann** [comment]
Yes, I will increase the idleness timeout to 5 minutes. Merging this PR then.
